browser/brave_ads/brave_ads_host.cc

```diff
@@ -0,0 +1,175 @@
+  brave_rewards::RewardsService* rewards_service =
+      brave_rewards::RewardsServiceFactory::GetForProfile(profile_);
+  if (!rewards_service || !ads_service || !ads_service->IsSupportedLocale()) {
+    std::move(callback).Run(false);
+    return;
+  }
+
+  if (ads_service->IsEnabled()) {
+    std::move(callback).Run(true);
+    return;
+  }
+
+  if (callback_) {
+    std::move(callback).Run(false);
+    return;
+  }
+
+  Browser* browser = chrome::FindBrowserWithWebContents(web_contents_);
+  if (!browser) {
+    std::move(callback).Run(false);
+    return;
+  }
+
+  callback_ = std::move(callback);
+
+  rewards_service->StartProcess(base::DoNothing());
+  rewards_service_observation_.Observe(rewards_service);
+  action_ui_observation_.Observe(extensions::BraveActionAPI::Get(browser));
+  initiated_browser_ = browser;
+
+  if (!ShowRewardsPopup(browser)) {
+    Reset();
+    std::move(callback_).Run(false);
+  }
+}
+
+void BraveAdsHost::WebContentsDestroyed() {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+
+  Reset();
+
+  if (callback_) {
+    std::move(callback_).Run(false);
+  }
+}
+
+void BraveAdsHost::OnRewardsEnabled() {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+
+  Reset();
+
+  if (!callback_) {
+    NOTREACHED();
+    return;
+  }
+
+  std::move(callback_).Run(true);
+}
+
+void BraveAdsHost::OnAdsEnabled(brave_rewards::RewardsService* rewards_service,
+                                bool ads_enabled) {
+  DCHECK(rewards_service);
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+
+  Reset();
+
+  if (!callback_) {
+    NOTREACHED();
+    return;
+  }
+
+  std::move(callback_).Run(ads_enabled);
+}
+
+void BraveAdsHost::OnActionUIClosed(Browser* browser,
+                                    const std::string& extension_id) {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+  if (extension_id != brave_rewards_extension_id) {
+    return;
+  }
+
+  const Browser* current_browser =
+      chrome::FindBrowserWithWebContents(web_contents_);
+  if (browser != current_browser && browser != initiated_browser_) {
+    return;
+  }
+
+  Reset();
+
+  if (!callback_) {
+    NOTREACHED();
+    return;
+  }
+
+  const AdsService* ads_service = AdsServiceFactory::GetForProfile(profile_);
+  bool ads_enabled = false;
+  if (ads_service) {
+    ads_enabled = ads_service->IsEnabled();
+  }
+  std::move(callback_).Run(ads_enabled);
+}
+
+bool BraveAdsHost::ShowRewardsPopup(Browser* browser) {
+  auto* extension_service =
+      extensions::ExtensionSystem::Get(profile_)->extension_service();
+  if (!extension_service) {
+    return false;
+  }
+
+  static_cast<extensions::BraveComponentLoader*>(
+      extension_service->component_loader())
+      ->AddRewardsExtension();
+
+  std::string error;
+  return extensions::BraveActionAPI::ShowActionUI(
+      browser, brave_rewards_extension_id,
+      std::make_unique<std::string>(kAdsEnableRelativeUrl), &error);
```

Comment:
woah, why does it take a `unique_ptr`??.. Not to be addressed by this PR but it doesn't look good

Reply:
I, er, have no idea why I would have done that. First month on c++ I guess...